src/lib/Microsoft.Health.Events/EventHubProcessor/PartitionLockingProcessor.cs

```diff
@@ -0,0 +1,418 @@
+// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure;
+using Azure.Core;
+using Azure.Identity;
+using Azure.Storage.Blobs;
+using Azure.Storage.Blobs.Models;
+
+// description
+// - Get the number of total processors and total partitions
+// - Calculate how many partitions should be assigned to each processor
+// - Lock the partition(s)
+// - Pass the partition ids into the StaticPartitionProcessor and run until the number of total processors changes
+
+// usage
+//  var cts = new CancellationTokenSource();
+//  var processor = new PartitionLockingProcessor();
+//  await processor.RunProcessor(cts);
+
+// todos:
+// - detect if a partition has no ownership updates for some time
+// - handle cases where the ratio of processors to partitions is not divided evenly
+// - detect if acquiring partitions takes too long
+// - replace Task.Delay with something more sophisticated
+
+namespace Microsoft.Health.Events.EventHubProcessor
+{
+    public class PartitionLockingProcessor
```

Comment:
The PartitionAssignment logic here seems to be separate from the actual EventProcessor/Service. What would you think about separating these processes out completely? I'm thinking of something like:

1. Run Partition Locking Processor as a Background Service inside of our Keda External Scaler.
  1. Responsible for maintaining ownership metadata on blob based on Active Processors
1. Run Loop within each service to maintain Active Processor blobs
1. Run StaticPartitionProcessor within each service.
   1. This periodically acquires the partitions to work on from blob storage, which is managed by the Partition Locking Processor running inside of Keda   

One downside to this approach is that if we scaled our services outside of Keda then this wouldn't work. 

Interested to hear your thoughts on this.

Reply:
It is a good suggestion. I am leaning towards running the partition assignment background service as part of the application code because we need a mechanism to restart the application loop if the number of processors change. We could create a separate service for that but for simplicity it is probably easier to include it as part of the partition assignment. I also like that it allows us to still scale using Keda or scale without using Keda